File: sayiir-nodejs/src/worker.ts
/**
 * Distributed worker for processing workflows across multiple processes.
 *
 * The Worker class polls a backend for available tasks, claims them,
 * and executes them using the registered task functions.
 *
 * **Note:** The distributed worker requires a PostgreSQL backend for
 * cross-process coordination. InMemoryBackend can be used for testing.
 *
 * @example
 * ```ts
 * const backend = PostgresBackend.connect(process.env.DATABASE_URL!);
 * const worker = new Worker("worker-1", backend, [orderFlow], {
 *   pollInterval: "5s",
 * });
 * const handle = worker.start();
 * process.on("SIGTERM", () => handle.shutdown());
 * ```
 */

import type { Duration, TaskCallback } from "./types.js";
import type { Workflow } from "./flow.js";
import {
  type Backend,
  InMemoryBackend,
  PostgresBackend,
} from "./executor.js";
import type { NapiWorker, NapiWorkerHandle } from "./native.js";
import { getNative } from "./native.js";
import { parseDuration } from "./duration.js";

/** Worker configuration options. */
export interface WorkerOptions {
  pollInterval?: Duration;
  claimTtl?: Duration;
  /** Affinity tags for this worker. When set, the worker only picks up tasks whose tags are a subset of these tags. */
  tags?: string[];
}

/** Handle for controlling a running worker. */
export class WorkerHandle {
  /** @internal */
  private readonly _native: NapiWorkerHandle;

  /** @internal */
  constructor(native: NapiWorkerHandle) {
    this._native = native;
  }

  /** Request a graceful shutdown. */
  shutdown(): void {
    this._native.shutdown();
  }
}

/** Distributed workflow worker. */
export class Worker {
  readonly workerId: string;
  readonly backend: Backend;
  readonly workflows: readonly Workflow<any, any>[];
  readonly options: WorkerOptions;

  constructor(
    workerId: string,
    backend: Backend,
    workflows: readonly Workflow<any, any>[],
    opts?: WorkerOptions,
  ) {
    this.workerId = workerId;
    this.backend = backend;
    this.workflows = workflows;
    this.options = opts ?? {};
  }

  /**
   * Start the worker and return a handle for lifecycle control.
   *
   * Spawns a background thread that polls for available tasks, claims them,
   * and dispatches them to the registered task functions.
   */
  start(): WorkerHandle {
    const napiWorker = this.createNapiWorker();

    // Build a combined task registry from all workflows
    const registry: Record<string, TaskCallback> = {};
    for (const wf of this.workflows) {
      Object.assign(registry, wf._taskRegistry);
    }

    // The native worker calls this dispatcher from a background thread.
    // It receives a JSON payload `{ taskId, input }` and must return
    // a JSON-serialized output string.
    const dispatcher = async (payload: string): Promise<string> => {
      const { taskId, input } = JSON.parse(payload) as {
        taskId: string;
        input: unknown;
      };
      const fn = registry[taskId];
      if (!fn) {
        throw new Error(`Task '${taskId}' not found in any workflow registry`);
      }
      const result = await fn(input);
      return JSON.stringify(result);
    };

    const napiHandle = napiWorker.start(
      this.workflows.map((wf) => wf._inner),
      dispatcher,
    );

    return new WorkerHandle(napiHandle);
  }

  /** @internal Create the native worker with the appropriate backend. */
  private createNapiWorker(): NapiWorker {
    const native = getNative();
    const pollMs = this.options.pollInterval
      ? parseDuration(this.options.pollInterval)
      : undefined;
    const claimMs = this.options.claimTtl
      ? parseDuration(this.options.claimTtl)
      : undefined;

    const tags = this.options.tags;

    if (this.backend instanceof InMemoryBackend) {
      return native.NapiWorker.withInMemory(
        this.workerId,
        this.backend._inner,
        pollMs,
        claimMs,
        tags,
      );
    }

    if (this.backend instanceof PostgresBackend) {
      return native.NapiWorker.withPostgres(
        this.workerId,
        this.backend._inner,
        pollMs,
        claimMs,
        tags,
      );
    }

    const backend = this.backend as unknown;
    const received =
      (backend as { constructor?: { name?: string } })?.constructor?.name ??
      typeof backend;
    throw new Error(
      `Unsupported backend type: ${received}. Expected InMemoryBackend or PostgresBackend.`,
    );
  }
}

File: sayiir-python/sayiir/worker.py
"""Distributed worker for processing workflows across multiple processes.

The Worker polls a backend for available tasks, claims them, and executes
them using the registered task functions. Multiple workers can run across
machines or processes, all polling the same backend.

Example::

    from sayiir import Flow, PostgresBackend, task
    from sayiir.worker import Worker

    @task
    def process_order(order):
        ...

    wf = Flow("orders").then(process_order).build()
    backend = PostgresBackend("postgresql://localhost/sayiir")
    worker = Worker("worker-1", backend)
    handle = worker.start([wf])

    # In shutdown handler:
    handle.shutdown()
    handle.join()
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ._sayiir import PyWorker as _PyWorker

if TYPE_CHECKING:
    from ._sayiir import PyWorkerHandle as _PyWorkerHandle
    from .flow import Workflow


class WorkerHandle:
    """Handle for controlling a running worker.

    Obtained from :meth:`Worker.start`. Use :meth:`shutdown` to request
    graceful shutdown and :meth:`join` to wait for completion.
    """

    def __init__(self, inner: _PyWorkerHandle) -> None:
        self._inner = inner

    def shutdown(self) -> None:
        """Request a graceful shutdown. Non-blocking."""
        self._inner.shutdown()

    def join(self) -> None:
        """Wait for the worker to finish. Call :meth:`shutdown` first.

        Releases the GIL while waiting.
        """
        self._inner.join()


class Worker:
    """Distributed workflow worker.

    Args:
        worker_id: Unique identifier for this worker node.
        backend: Either ``InMemoryBackend()`` or ``PostgresBackend(url)``.
        poll_interval: Seconds between polls for available tasks.
        claim_ttl: Task claim TTL in seconds.
    """

    def __init__(
        self,
        worker_id: str,
        backend: Any,
        *,
        poll_interval: float = 5.0,
        claim_ttl: float = 300.0,
    ) -> None:
        self._inner = _PyWorker(worker_id, backend, poll_interval, claim_ttl)

    def start(self, workflows: list[Workflow]) -> WorkerHandle:
        """Start the worker and return a handle for lifecycle control.

        Args:
            workflows: List of compiled :class:`~sayiir.Workflow` objects.
                Each workflow's task registry is used to look up task
                functions when executing claimed tasks.

        Returns:
            A :class:`WorkerHandle` for shutdown and lifecycle control.
        """
        pairs = [(wf._inner, wf._task_registry) for wf in workflows]
        return WorkerHandle(self._inner.start(pairs))
